File: julia/14_contraction.jl
#!/usr/bin/env julia
"""
Tensor contraction examples
"""

using Random
using TensorOperations

function main()
    println("matrix-matrix multiplication")
    A = rand(2, 3)
    B = rand(3, 4)
    println("A: shape ", size(A), "\n", A, "\n")
    println("B: shape ", size(B), "\n", B, "\n")
    println("contract: A, B -> C")
    @tensor C[i, k] := A[i, j] * B[j, k]
    println("C: shape ", size(C), "\n", C, "\n")

    println("more complex contraction")
    A = rand(2, 3, 4, 5)
    B = rand(4, 3)
    C = rand(5, 3, 4)
    println("A: shape ", size(A), "\n", A, "\n")
    println("B: shape ", size(B), "\n", B, "\n")
    println("C: shape ", size(C), "\n", C, "\n")
    println("contract: A, B, C -> D")
    @tensor D[i, j, k] := A[i, j, l, m] * B[l, n] * C[m, n, k]
    println("D: shape ", size(D), "\n", D)
end

main()

File: python/14_contraction.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Tensor contraction examples
"""

import numpy as np


def main():
    print("matrix-matrix multiplication")
    A = np.random.random((2, 3))
    B = np.random.random((3, 4))
    print(f"A: shape {A.shape}\n{A}\n")
    print(f"B: shape {B.shape}\n{B}\n")
    print("contract: A, B -> C")
    C = np.einsum("ij,jk->ik", A, B)
    print(f"C: shape {C.shape}\n{C}\n")

    print("more complex contraction")
    A = np.random.random((2, 3, 4, 5))
    B = np.random.random((4, 3))
    C = np.random.random((5, 3, 4))
    print(f"A: shape {A.shape}\n{A}\n")
    print(f"B: shape {B.shape}\n{B}\n")
    print(f"C: shape {C.shape}\n{C}\n")
    print("contract: A, B, C -> D")
    D = np.einsum("ijlm,ln,mnk->ijk", A, B, C)
    print(f"D: shape {D.shape}\n{D}")


if __name__ == "__main__":
    main()
